### ISSUE TITLE
TAIT: it's possible to impl a trait for a tait by using projections

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#99840]
```rust
#![feature(type_alias_impl_trait)]
type Alias = impl Sized;

fn constrain() -> Alias {
    1i32
}

trait HideIt {
    type Assoc;
}

impl HideIt for () {
    type Assoc = Alias;
}

pub trait Yay {}

impl Yay for <() as HideIt>::Assoc {}
// impl Yay for i32 {} // this errors
// impl Yay for u32 {} // this also errors
```
As coherence checking keeps opaque types opaque this should not be unsound. Considering that a direct impl for opaque types is forbidden, this is an inconsistency we should resolve though.

cc @oli-obk

**Comment by oli-obk** [comment]
The two commented out lines do not cause any errors (https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8366951aa5afd218745dd0ebc4564387) So I'd say this is unsound or at least causes ICEs: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=07fbbfb3b0ecc1dc59ffc71b18b06f1e

**Comment by lcnr** [comment]
what, is that a recent regression? :o these do error with `rustc 1.63.0-nightly (5435ed691 2022-06-07)`

**Comment by lqd** [comment]
It looks to be from last week's https://github.com/rust-lang/rust/pull/99506, and out of that rollup [#99345](https://github.com/rust-lang/rust/pull/99345) looks likely.

**Comment by oli-obk** [comment]
Hmm... that one got reverted in https://github.com/rust-lang/rust/pull/99079

cc @compiler-errors

**Comment by lqd** [comment]
Was that a revert, or a targeted fix for closures ?

**Comment by oli-obk** [comment]
revert + actual fix of the original issue

**Comment by lqd** [comment]
The behavior is the same on master, no errors + ICEs. I'll try to manually bisect within the rollup to see confirm cargo-bisect-rustc's findings.

**Comment by compiler-errors** [comment]
Isn't this just another form of #99554?

**Comment by lqd** [comment]
(I wish we had try builds for each rolled up PR 😅) It was indeed introduced in the same rollup: by eecfdfb8a12367cb292a85b6a30981eb35a56379 / #99383.

Reverting *that* does indeed bring the conflicting implementation errors back, and fixes the ICE.

**Comment by lcnr** [comment]
> Isn't this just another form of #99554?

no, #99554 is using projections in ways which should fail the orphan check, here we have projections which should pass the orphan check, but can be normalized to opaque types.

These issues need two distinct fixes

**Comment by oli-obk** [comment]
I have a proper fix locally, but we should investigate how #99383 affected this

**Comment by ouz-a** [comment]
> (I wish we had try builds for each rolled up PR 😅) It was indeed introduced in the same rollup: by [eecfdfb](https://github.com/rust-lang/rust/commit/eecfdfb8a12367cb292a85b6a30981eb35a56379) / #99383.
> 
> Reverting _that_ does indeed bring the conflicting implementation errors back, and fixes the ICE.

Although it looks suspicious this is not caused by #99383 I reverted all changes made by it and tested it, this still passes without any errors. Here https://github.com/ouz-a/rust/tree/undo_formalize
